src/adapters/bridge-api.ts

```diff
@@ -0,0 +1,74 @@
+import axios from "axios";
+import { z } from "zod";
+
+import { StrictSchema } from "src/utils/type-safety";
+import * as domain from "src/domain";
+
+interface Bridge {
+  token_addr: string;
+  amount: string;
+  dest_net: number;
+  dest_addr: string;
+  deposit_cnt: string;
+}
+
+interface GetBridgesParams {
+  env: domain.Env;
+  ethereumAddress: string;
+}
+
+interface GetBridgesResponse {
+  deposits?: Bridge[];
+}
+
+const bridgeParser = StrictSchema<Bridge>()(
+  z.object({
+    token_addr: z.string(),
+    amount: z.string(),
+    dest_net: z.number(),
+    dest_addr: z.string(),
+    deposit_cnt: z.string(),
+  })
+);
+
+const getBridgesResponseParser = StrictSchema<GetBridgesResponse>()(
+  z.object({
+    deposits: z.optional(z.array(bridgeParser)),
+  })
+);
+
+const transformApiBridgeToDomain = ({
+  token_addr,
+  amount,
+  dest_addr,
+  dest_net,
+  deposit_cnt,
+}: Bridge): domain.Bridge => ({
+  tokenAddress: token_addr,
+  amount: amount,
+  destinationAddress: dest_addr,
+  destinationNetwork: dest_net,
+  depositCount: deposit_cnt,
+});
```

Comment:
Instead of using this _Persistence to Domain_ transformation function as an external artifact, I'd rather use [Zod's built in transformation](https://github.com/colinhacks/zod#transform) mechanism:

```typescript
const bridgeParser = StrictSchema<domain.Bridge, Bridge>()(
  z
    .object({
      token_addr: z.string(),
      amount: z.string(),
      dest_net: z.number(),
      dest_addr: z.string(),
      deposit_cnt: z.string(),
    })
    .transform(({ token_addr, amount, dest_addr, dest_net, deposit_cnt }) => ({
      tokenAddress: token_addr,
      amount: amount,
      destinationAddress: dest_addr,
      destinationNetwork: dest_net,
      depositCount: deposit_cnt,
    }))
);
```

Like this we rely on the parser itself to transform the data to our domain, making sure that whenever we use it to decode the API input, we get our domain entity as output.

Note that for this to work we need to improve a bit our `StrictSchema` helper to accept two types, one as output and a new one, which is optional, as input: `StrictSchema<domain.Bridge, Bridge>`:

```typescript
export const StrictSchema: <O, I = O>() => <UO, UI = UO>(
  u: Exact<ZodSchema<O, ZodTypeDef, I>, ZodSchema<UO, ZodTypeDef, UI>> extends true
    ? Exact<Required<O>, Required<UO>> extends true
      ? ZodSchema<UO, ZodTypeDef, UI>
      : never
    : never
) => ZodSchema<O, ZodTypeDef, I> =
  () =>
  <O, I = O>(u: unknown) =>
    u as ZodSchema<O, ZodTypeDef, I>;
```

Like this, we optionally accept an _Input_ type parameter which, as Zod does, defaults to the already required _Output_ type parameter and provide it to the `ZodSchema`.

Reply:
The only thing that I don't like of this approach is that is specific of `zod` and it's not portable, while the other is just a `map` and a function (which is portable to any parsing library). Which could be suitable for keeping it reusable is keeping the function and doing this:

```ts
const bridgeParser = StrictSchema<domain.Bridge, Bridge>()(
  z
    .object({
      token_addr: z.string(),
      amount: z.string(),
      dest_net: z.number(),
      dest_addr: z.string(),
      deposit_cnt: z.string(),
    })
    .transform(transformApiBridgeToDomain)
);
```

Are you okay with this approach?

Reply:
Of course! The important thing is that when you use the parser you get the domain entity as output.

I'd though remove the prefix `transform` from the function name, so it reads `.transform(apiBridgeToDomain)`.

The name `apiBridgeToDomain` already implies a transformation.